tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java

```diff
@@ -0,0 +1,256 @@
+import java.nio.FloatBuffer;
+
+import uk.ac.manchester.tornado.api.internal.annotations.SegmentElementSize;
+import uk.ac.manchester.tornado.api.types.HalfFloat;
+import uk.ac.manchester.tornado.api.types.arrays.TornadoNativeArray;
+
+@SegmentElementSize(size = 4) //This needs to be fixed
+public final class Tensor extends TornadoNativeArray implements AbstractTensor {
+
+    /**
+     * The data type of the elements contained within the tensor.
+     */
+    private final DType dType;
+
+    /**
+     * The total number of elements in the tensor.
+     */
+    private int numberOfElements;
+
+    /**
+     * The memory segment representing the tensor data in native memory.
+     */
+    private MemorySegment segment;
+
+    /**
+     * The index offset in the memory segment where the tensor data starts.
+     */
+    private int baseIndex;
+
+    /**
+     * The size in bytes of the memory segment including the header.
+     */
+    private long segmentByteSize;
+
+    /**
+     * The shape of the tensor, representing its dimensions.
+     */
+    private Shape shape;
+
+    /**
+     * Constructs a tensor with a specified size, memory segment, and data type.
+     *
+     * @param size
+     *     the number of elements in the tensor
+     * @param memorySegment
+     *     the memory segment where the tensor is stored
+     * @param dtype
+     *     the data type of the tensor elements
+     */
+    // This is problematic, it needs the segment to be shifted to cater the ARRAY_HEADER
+    public Tensor(int size, MemorySegment memorySegment, DType dtype) {
+        assert size >= 0;
+        this.numberOfElements = size;
+        this.segment = memorySegment;
+        this.dType = dtype;
+    }
+
+    public Tensor(Shape shape, MemorySegment memorySegment, DType dtype) {
+        this.segment = memorySegment;
+        this.dType = dtype;
+        this.shape = shape;
+    }
+
+    /**
+     * Constructs a tensor with a specified shape and data type. The memory for the tensor is automatically allocated.
+     *
+     * @param shape
+     *     the shape of the tensor
+     * @param dtype
+     *     the data type of the tensor elements
+     */
+    public Tensor(Shape shape, DType dtype) {
+        this.shape = shape;
+        this.numberOfElements = shape.getSize();
+        this.dType = dtype;
+        this.baseIndex = (int) TornadoNativeArray.ARRAY_HEADER / dtype.getByteSize();
+        this.segmentByteSize = (long) numberOfElements * dtype.getByteSize() + TornadoNativeArray.ARRAY_HEADER;
+        this.segment = Arena.ofAuto().allocate(segmentByteSize, 1);
+        segment.setAtIndex(JAVA_INT, 0, numberOfElements);
+    }
+
+    //    TensorFP16=NEW\\\
+
+    /**
+     * Creates a tensor from a float array.
+     *
+     * @param values
+     *     the float array to convert into a tensor
+     * @return a new tensor containing the provided values
+     */
+    public static Tensor fromArray(float[] values) {
```

Comment:
There is only a static method to build a tensor from a `float` array. What if the input is an `int[]` array?